### ISSUE TITLE
IPIP encapsulation not working with eBPF on CentOS/AlamLinux 8.5

### ISSUE THREAD

**Issue by cristicalin** [projectcalico/calico#5449]
## Expected Behavior

I configured a test cluster of 3 nodes running AlmaLinux 8.5 with [kubespray](https://github.com/kubernetes-sigs/kubespray.git) with the default configuration (kubernetes 1.23.1, calico 3.21.2, ipipMode: Always) and deployed netchecker to validate configuration (netchecker deployment is part of kubespray). I expected to have connectivity between pods and the netchecker server to be reachable from the netchecker agent pods.

Netchecker is a service running a server in 1 pod exposed through a service on http (tcp port 8081) and two agent pods running on each node as two daemonsets, one pod is running on kubernetes network and one is running on the host network, these pods connect periodically through the kubernetes service to the server pod and check in to the server to account for successful network connectivity both between pods and from host to pods.

## Current Behavior

The netchecker service did not register any node and the agent pods were complaining that they could not reach the server. Log excerpt below:

Pod in the kubernetes network:
```
E0115 15:56:08.260819       1 agent.go:135] DNS look up host error. Details: lookup netchecker-service on 169.254.25.10:53: server misbehaving
I0115 15:56:08.261120       1 agent.go:114] Send payload via URL: http://netchecker-service:8081/api/v1/agents/netchecker-agent-chgvm
E0115 15:56:22.261328       1 agent.go:306] Error while sending info. Details: Post http://netchecker-service:8081/api/v1/agents/netchecker-agent-chgvm: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I0115 15:56:22.261349       1 agent.go:301] Sleep for 15 second(s)
```

Pod in the host namespace:
```
E0115 15:56:36.519065       1 agent.go:199] Get http://netchecker-service:8081/api/v1/ping: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
I0115 15:56:36.519138       1 agent.go:232] HTTP Probe (http://netchecker-service:8081/api/v1/ping): HTTPCode: 0; Total: 9223372036854 ms; ContentTransfer: 0 ms; Connect: 0 ms; DNSLookup: 0 ms; ServerProcessing: 0 ms; TCPConnection: 0 ms; 
E0115 15:57:07.523238       1 agent.go:135] DNS look up host error. Details: lookup netchecker-service on 169.254.25.10:53: server misbehaving
I0115 15:57:07.523493       1 agent.go:114] Send payload via URL: http://netchecker-service:8081/api/v1/agents/netchecker-agent-hostnet-5c2r2
```

`exec`'ing into an agent pod and trying to do a direct connection to the service IP got the `wget` tool stuck opening the tcp connection. This connection only seems to work on the same node, pods on the same node can communicate but not pods on different nodes.

```
(venv) [root@instance-1 kubespray]# /usr/local/bin/kubectl exec -it netchecker-agent-hostnet-5c2r2 -- sh
/ $ wget -Oq - 10.233.52.229:8081/api/v1/agents/
```

## Steps to Reproduce (for bugs)

1. Install 3 VMs with vagrant on the same physical host, vm OS should be Alma 8.5 or CentOS 8.5, with the 8.4 version this test works both on Alma and on CentOS
2. Deploy kubernetes with kubespray (https://github.com/kubernetes-sigs/kubespray/commit/c0d1bb1a5c988c3d0880060089ac39bdbf52f67a) with default settings plus:

```yaml
calico_bpf_enabled: true
calico_version: v3.21.2
ipip: true
loadbalancer_apiserver_localhost: true
use_localhost_as_kubeapi_loadbalancer: true
auto_renew_certificates: true
deploy_netchecker: true
dns_min_replicas: 1
```

3. Wait for the cluster to be deployed
4. Get service ip of `netchecker-service` and check the list of agents

```ShellSession
(venv) [root@instance-1 kubespray]# /usr/local/bin/kubectl get svc
NAME                 TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)          AGE
kubernetes           ClusterIP   10.233.0.1      <none>        443/TCP          28m
netchecker-service   NodePort    10.233.52.229   <none>        8081:31081/TCP   25m
(venv) [root@instance-1 kubespray]# curl -s 10.233.52.229:8081/api/v1/agents/ | jq .
{}
(venv) [root@instance-1 kubespray]# curl -s 10.233.52.229:8081/api/v1/connectivity_check | jq .
{
  "Message": "Connectivity check fails. Reason: there are absent or outdated pods; look up the payload",
  "Absent": [
    "netchecker-agent-chgvm",
    "netchecker-agent-hostnet-5c2r2",
    "netchecker-agent-hostnet-9lc7h",
    "netchecker-agent-hostnet-qw84x",
    "netchecker-agent-s2vsg",
    "netchecker-agent-sssj8"
  ],
  "Outdated": null
}
```

5. Expected result is a complex json with listing of agents and the connectivity check showing no absent nodes.

## Context

CI image upgrade in kubespray project showed this failure for the calico ebpf test case.

## Your Environment

* Calico version: v3.21.2
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes 1.23.1
* Operating System and version: AlmaLinux 8.5 / CentOS 8.5
* Link to your project (optional): https://github.com/kubernetes-sigs/kubespray.git

## Node details
https://gist.github.com/cristicalin/d910c6c9ba39a42042e6971f81bbd897

**Comment by cristicalin** [comment]
[bpf_dump.log.gz](https://github.com/projectcalico/calico/files/7875363/bpf_dump.log.gz)

The output from a `tc exec bpf debug` after setting `bpfLogLevel: "Debug"` in `felixconfig`

**Comment by tomastigera** [comment]
This is where the things are broken. We get an IPIP packet, but by the time it gets to the IPIP device, it is too short.

This is likely not true in your configuration https://github.com/projectcalico/calico/blob/master/felix/bpf-gpl/skb.h#L70-L72

```
    tmux: server-2794  [000] d.s.  3220.016223: bpf_trace_printk: ens5-----I: New packet at ifindex=2; mark=0
    tmux: server-2794  [000] d.s.  3220.016223: bpf_trace_printk: ens5-----I: No metadata is shared by XDP
    tmux: server-2794  [000] d.s.  3220.016223: bpf_trace_printk: ens5-----I: IP id=7210 s=c0a87927 d=c0a87959
    tmux: server-2794  [000] d.s.  3220.016224: bpf_trace_printk: ens5-----I: IPIP packet from known Calico host, allow.
    tmux: server-2794  [000] d.s.  3220.016225: bpf_trace_printk: ens5-----I: Traffic is towards host namespace, marking with 1000000.
    tmux: server-2794  [000] d.s.  3220.016225: bpf_trace_printk: ens5-----I: Final result=ALLOW (0). Program execution time: 1980ns


    tmux: server-2794  [000] d.s.  3220.016236: bpf_trace_printk: tunl0----I: New packet at ifindex=3; mark=1000000
    tmux: server-2794  [000] d.s.  3220.016237: bpf_trace_printk: tunl0----I: No metadata is shared by XDP
    tmux: server-2794  [000] d.s.  3220.016237: bpf_trace_printk: tunl0----I: Pulling 62 bytes.
    tmux: server-2794  [000] d.s.  3220.016238: bpf_trace_printk: tunl0----I: Pull failed (min len)
    tmux: server-2794  [000] d.s.  3220.016238: bpf_trace_printk: tunl0----I: Too short
    tmux: server-2794  [000] d.s.  3220.016238: bpf_trace_printk: tunl0----I: Drop malformed or unsupported packet
    tmux: server-2794  [000] d.s.  3220.016239: bpf_trace_printk: tunl0----I: Final result=DENY (1). Program execution time: 1264ns
```

**Comment by fasaxc** [comment]
IIRC, IPIP is a L3 device so we expect packets to arrive with no ethernet header.  Do we have a hard-coded assumption about the min length that assumes an ethernet header?

**Comment by mazdakn** [comment]
@cristicalin can you check if the HW offload is enabled in your environment? and possibly what happens if it is turned off?

**Comment by mazdakn** [comment]
@cristicalin we identified the issue is caused by a change in kernel that impacts how IPIP tunnels work internally. I am working on a fix for it: https://github.com/projectcalico/calico/pull/5846

**Comment by mazdakn** [comment]
The fix will be available in Calico 3.23.0
